File: src_local/agents/Support/markdown_sanitizer.py
"""
📝 Sanitizador de Markdown (MarkdownSanitizer).
Realiza a limpeza e normalização garantindo conformidade absoluta com MD041 e MD047.
"""
import re
import logging
import time

logger = logging.getLogger(__name__)

class MarkdownSanitizer:
    def sanitize(self, content: str) -> str:
        """Soberania: Limpeza PhD e normalização estratégica de Markdown."""
        if not content: return ""
        start_time = time.time()
        
        content = content.lstrip('\ufeff').strip()
        lines = self._process_structure(content.split('\n'))
        lines = self._ensure_h1(lines)
        
        sanitized = '\n'.join(lines).strip() + '\n'
        
        from src_local.utils.logging_config import log_performance
        log_performance(logger, start_time, "📝 Markdown Sanitized (PHD Core)")
        return sanitized

    def _process_structure(self, raw_lines: list) -> list:
        """Processa blocos de código, headers e espaçamento vertical."""
        res, seen, in_cb = [], {}, False
        
        for i, line in enumerate(raw_lines):
            stripped = line.rstrip()
            in_cb, processed = self._handle_block_logic(stripped, in_cb)
            if processed:
                res.append(processed)
                continue

            if re.match(r'^\#+\s+', stripped.strip()):
                self._process_header(stripped.strip(), res, seen, raw_lines, i)
            elif stripped.strip() or (res and res[-1]):
                res.append(stripped)
                
        # Remove linhas vazias do início (MD041)
        while res and not res[0].strip():
            res.pop(0)
            
        return res

    def _handle_block_logic(self, line: str, in_cb: bool):
        """Gerencia estado de code blocks."""
        stripped = line.strip()
        if stripped.startswith('```'):
            return not in_cb, line
        if in_cb:
            return True, line
        return False, None

    def _process_header(self, line: str, res: list, seen: dict, raw: list, idx: int):
        """Normaliza cabeçalhos e aplica regras MD022/MD026."""
        if res and res[-1]: res.append('')
        
        h = re.sub(r'[\.\!\?\:]+$', '', line)
        if h in seen:
            seen[h] += 1
            h = f"{h} [v{seen[h]}]"
        else: seen[h] = 1
        
        res.append(h)
        # Padding Below
        if idx < len(raw)-1 and raw[idx+1].strip(): res.append('')

    def _ensure_h1(self, lines: list) -> list:
        """Garante a presença de um H1 no topo do documento."""
        if not lines: return ["# 🏛️ RELATÓRIO DE CONSCIÊNCIA SISTÊMICA", ""]
        
        # Check first line
        if lines[0].strip().startswith('# '): return lines
        
        # Search for buried H1
        for idx, line in enumerate(lines):
            if line.strip().startswith('# '):
                h1 = lines.pop(idx)
                return [h1, ""] + lines
                
        # No H1 found
        return ["# 🏛️ RELATÓRIO DE CONSCIÊNCIA SISTÊMICA", ""] + lines

File: src_local/agents/Support/Reporting/markdown_sanitizer.ts
import winston from "winston";
import { MarkdownStructureAgent } from "./markdown_structure_agent.ts";

const logger = winston.child({ module: "MarkdownSanitizer" });

/**
 * 📝 Sanitizador de Markdown (High-Fidelity TypeScript Version).
 * Realiza a limpeza e normalização garantindo conformidade absoluta com MD041 e MD047.
 * 
 * Melhorias sobre a versão legacy:
 * 1. Performance otimizada usando manipulação de strings Bun.
 * 2. Proteção contra BOM (Byte Order Mark).
 * 3. Fallback inteligente para relatórios sem título.
 * 4. [PhD Update] Suporte a Markdown aninhado (Blockquotes) para MD022 e MD032.
 */
export class MarkdownSanitizer {
    /**
     * Sanitiza o conteúdo de um arquivo Markdown.
     * @param content Conteúdo bruto do markdown.
     */
    public async sanitize(content: string): Promise<string> {
        if (!content) return "";
        const startTime = Date.now();

        // 1. Limpeza de caracteres especiais (como BOM) e normalização de quebra de linha
        let internalContent = content.replace(/^\ufeff+/, '').replace(/\r\n/g, "\n").trim();
        let lines = internalContent.split("\n");

        // 2. Delegação para o Agente de Estrutura para refinamento semântico
        const structureAgent = new MarkdownStructureAgent(lines);
        lines = structureAgent.executeRefinement();

        // 3. Aplicação de políticas de conformidade
        lines = this.cleanup(lines);
        lines = this.ensureBlanksAroundHeadings(lines); // MD022
        lines = this.ensureBlanksAroundLists(lines);    // MD032
        lines = this.ensureH1(lines);

        // 4. Junção com garantia de newline no final (MD047)
        const sanitized = lines.join("\n").trim() + "\n";

        const duration = Date.now() - startTime;
        logger.info(`📝 Markdown Sanitized em ${duration}ms (Incluso refinamento estrutural)`);

        return sanitized;
    }

    private cleanup(lines: string[]): string[] {
        // Remove linhas vazias no início e no fim
        while (lines.length > 0 && lines[0]!.trim().length === 0) lines.shift();
        while (lines.length > 0 && lines[lines.length - 1]!.trim().length === 0) lines.pop();
        return lines;
    }

    private ensureH1(lines: string[]): string[] {
        // Se vazio, cria um cabeçalho padrão
        if (lines.length === 0) return ["# 🏛️ RELATÓRIO SISTÊMICO", ""];

        // Se a primeira linha já é H1, está correto
        if (lines[0]!.trim().startsWith("# ")) return lines;

        // Procura por um H1 em outras partes e o promove para o topo
        for (let i = 0; i < lines.length; i++) {
            if (lines[i]!.trim().startsWith("# ")) {
                const [h1] = lines.splice(i, 1);
                return [h1!, "", ...lines];
            }
        }

        // Se não encontrar nenhum H1, cria um
        return ["# 🏛️ RELATÓRIO SISTÊMICO", "", ...lines];
    }

    private ensureBlanksAroundHeadings(lines: string[]): string[] {
        const result: string[] = [];
        for (let i = 0; i < lines.length; i++) {
            const line = lines[i]!;
            const clean = line.replace(/^>\s*/, "").trim();
            const prefix = line.match(/^>\s*/)?.[0] || "";
            const isHeading = clean.startsWith("#");

            if (isHeading) {
                // Ensure blank line before (unless first line)
                if (result.length > 0) {
                    const prev = result[result.length - 1]!;
                    const prevClean = prev.replace(/^>\s*/, "").trim();
                    if (prevClean !== "") {
                        result.push(prefix.trimEnd());
                    }
                }
                result.push(line);

                // Ensure blank line after
                if (i + 1 < lines.length) {
                    const next = lines[i + 1]!;
                    const nextClean = next.replace(/^>\s*/, "").trim();
                    if (nextClean !== "") {
                        result.push(prefix.trimEnd());
                    }
                }
            } else {
                result.push(line);
            }
        }
        return result;
    }

    private ensureBlanksAroundLists(lines: string[]): string[] {
        const result: string[] = [];
        const isList = (l: string) => /^\s*([-*]|\d+\.)\s+/.test(l.replace(/^>\s*/, ""));
        const isHorizontalRule = (l: string) => /^\s*-{3,}\s*$/.test(l.replace(/^>\s*/, ""));

        for (let i = 0; i < lines.length; i++) {
            const line = lines[i]!;
            const prefix = line.match(/^>\s*/)?.[0] || "";
            const prev = i > 0 ? lines[i - 1]! : "";
            const clean = line.replace(/^>\s*/, "").trim();

            // MD032: Blank line BEFORE list
            const prevClean = prev.replace(/^>\s*/, "").trim();
            if (isList(line) && !isList(prev) && prevClean !== "") {
                if (result.length > 0 && result[result.length - 1]!.replace(/^>\s*/, "").trim() !== "") {
                    result.push(prefix.trimEnd());
                }
            }

            result.push(line);

            // MD032: Blank line AFTER list
            const next = i + 1 < lines.length ? lines[i + 1]! : "";
            const nextClean = next.replace(/^>\s*/, "").trim();

            if (isList(line) && !isList(next) && (nextClean !== "" || isHorizontalRule(next))) {
                if (nextClean !== "") {
                    result.push(prefix.trimEnd());
                }
            }
        }
        return result;
    }
}

// Extensão utilitária para strings similar ao Python lstrip
declare global {
    interface String {
        lstrip(chars?: string): string;
    }
}

if (!String.prototype.lstrip) {
    String.prototype.lstrip = function (chars?: string): string {
        if (!chars) return this.trimStart();
        let start = 0;
        while (start < this.length && chars.includes(this[start]!)) {
            start++;
        }
        return this.substring(start);
    };
}
